### DISCUSSION TITLE
Jest is unable to consume `mjs` files imported by `@nuxt/bridge` via `#app`

### DISCUSSION THREAD

**Discussion by sarayourfriend** [nuxt/nuxt#18122] [Questions]
### Environment

- Operating System: `Darwin`
- Node Version:     `v14.17.5`
- Nuxt Version:     `2.16.0-27226092.034b9901`
- Package Manager:  `npm`
- Bundler:          `Webpack`
- User Config:      `version`, `cache`, `srcDir`, `modern`, `server`, `router`, `components`, `plugins`, `css`, `head`, `env`, `dev`, `buildModules`, `styleResources`, `modules`, `i18n`, `sentry`, `tailwindcss`, `storybook`
- Runtime Modules:  `@nuxtjs/sentry@5.0.3`, `@nuxtjs/sitemap@2.4.0`, `nuxt-ssr-cache@1.5.2`, `@nuxtjs/i18n@7.1.0`
- Build Modules:    `@nuxtjs/tailwindcss@4.2.1`, `@nuxtjs/style-resources@1.0.0`, `@nuxtjs/svg@0.1.12`, `@nuxtjs/eslint-module@3.0.2`, `@nuxt/bridge@3.0.0-27246961.9d40a27`

### Describe the bug

Jest is unable to resolve the `#app` import by default. Adding a `moduleNameMapper` and a `transform` option for `mjs` helps fix initial problems but does not ultimately resolve the issue:

```js
module.exports = {
  moduleFileExtensions: ['js', 'ts', 'vue', 'json', 'mjs'],
  moduleNameMapper: {
    '^#app$': '@nuxt/bridge/dist/runtime/index.mjs',
    '^@/(.*)$': '<rootDir>/src/$1',
    '^~/(.*)$': '<rootDir>/src/$1',
    '^vue$': 'vue/dist/vue.common.js',
    '(.*svg)(\\?inline)$': '<rootDir>/test/unit/test-utils/svgTransform.js',
  },
  transform: {
    '^.+\\.m?[tj]s$': 'babel-jest',
    '.*\\.(vue)$': 'vue-jest',
    '.+\\.(css|styl|less|sass|scss|png|jpg|ttf|woff|woff2)$':
      'jest-transform-stub',
    '^.+\\.svg$': '<rootDir>/test/unit/svg-transform.js',
  },
}
```

But it doesn't ultimately fix the issue. Jest will still fail like so:

```
 FAIL  pages/index.test.js
  ● Test suite failed to run

    Jest encountered an unexpected token

    This usually means that you are trying to import a file which Jest cannot parse, e.g. it's not plain JavaScript.

    By default, if Jest sees a Babel config, it will use that to transform your files, ignoring "node_modules".

    Here's what you can do:
     • If you are trying to use ECMAScript Modules, see https://jestjs.io/docs/en/ecmascript-modules for how to enable it.
     • To have some of your "node_modules" files transformed, you can specify a custom "transformIgnorePatterns" in your config.
     • If you need a custom transformation specify a "transform" option in your config.
     • If you simply want to mock your non-JS modules (e.g. binary assets) you can stub them out with the "moduleNameMapper" config option.

    You'll find more details and examples of these config options in the docs:
    https://jestjs.io/docs/en/configuration.html

    Details:

    /sandbox/node_modules/@nuxt/bridge/dist/runtime/index.mjs:1
    ({"Object.<anonymous>":function(module,exports,require,__dirname,__filename,global,jest){export * from './app.mjs'
                                  ^^^^^^

    SyntaxError: Unexpected token 'export'

      4 |
      5 | <script>
    > 6 | import { ref } from "#app";
        | ^
      7 | export default {
      8 |   setup() {
      9 |     const foo = ref(0);

      at Runtime.createScriptFromCode (node_modules/jest-runtime/build/index.js:1350:14)
      at pages/index.vue:6:1
      at Object.<anonymous> (pages/index.vue:25:3)
```

`@nuxt/bridge` could publish regular `js` files to be used by `jest`. Or else some advice for how to resolve this and make it possible for `jest` to run with `@nuxt/bridge` would be very helpful and welcome.[](url)

### Reproduction

https://codesandbox.io/s/adoring-moore-q87xq

Open a new terminal in that code sandbox and run `yarn test` and observe the error related to importing `@nuxt/bridge`

**Comment by FGoessler** [comment]
I am having the exact same issue with plain Nuxt3. Would be in general great if more support for testing in the docs and the general tooling would be added because without testing it's not really possible to build quality stuff on top of Nuxt3! :)

**Comment by FGoessler** [comment]
Hi @sarayourfriend,

I got some steps further in my problem and this might help you as well:

For me the issue was that `.mjs` files of the nuxt3 node_modules weren't transformed by babel-jest. I then  learned that in my case this was due to the fact that I used a `.babelrc` file to configure babel. However `.babelrc` files do not apply anymore to `node_modules`. If you want to configure that you have to use a `babel.config.js` file for your project instead.

In my case as simple as:
```
// babel.config.js
module.exports = {
  "presets": ["@babel/env", "@babel/preset-typescript"]
}
```

Of course you have to keep `transformIgnorePatterns ` in your `jest.config.js` - in my case for nuxt3:
```
  ...
  transformIgnorePatterns: [
    "node_modules/(?!(nuxt3|unenv))",
  ]
  ...
```

Source re `.babelrc` vs `babel.config.js`: https://github.com/facebook/jest/issues/6053#issuecomment-383632515

**Reply by sarayourfriend** [reply]
Thank you so much @FGoessler! That is super helpful. When we have time on Openverse to revisit Nuxt 3 I will definitely keep this in mind. Much appreciated 🙂

**Reply by FGoessler** [reply]
Am happy this helped! 🙌

I now also had the time to finish the blogpost about my setup: https://medium.com/@fgoessler/set-up-nuxt3-with-jest-typescript-80aa4d3cfabc
